Add the required Golang tests for the ECAA.Registrar feature

<!-- For general purpose questions, use Stack Overflow http://stackoverflow.com/questions/tagged/hyperledger -->
## Description

Following @dulcep's #892 "Enable dynamic user registration" and the various discussions around #1186  regarding the security aspects around hard-coded username & password pairs vs. being able to have the various registrar privilege role to secure the  `ECA.RegisterUser`. 

The 89 files changed/added in PR #1723 were successfully merged to master on Friday (June 9, 12:06 PDT), and included 7 newly introduced functions added to `membersrvc/ca/ca.go`.

In order to help with system testing, integration tests, and other good reasons, we let the merge go through. Just so that we don't forget: these newly added functions were merged with 0 Golang tests.
## Describe How to Reproduce

<!-- If an issue, provide sufficient context and steps to reproduce the issue -->

Run `go test` with coverage flags and see the below coverage levels:

```
fabric/membersrvc/ca/ca.go:725:          canRegister                0.0%
fabric/membersrvc/ca/ca.go:757:          newMemberMetadata          0.0%
fabric/membersrvc/ca/ca.go:784:          canRegister                0.0%
fabric/membersrvc/ca/ca.go:801:          strsContained              0.0%
fabric/membersrvc/ca/ca.go:814:          strContained               0.0%
fabric/membersrvc/ca/ca.go:824:          role2String                0.0%
...
```

For some perspective, here a snippet of other existing functions in the vicinity (and many thanks for all those who help me/us daily to sort out these various `membersrvc` items, clearly, it helps a lot):

```
...
fabric/membersrvc/ca/aca.go:551:  startACAP                 100.0%
fabric/membersrvc/ca/aca.go:556:  Start                     100.0%
fabric/membersrvc/ca/ca.go:72:    NewCertificateSpec        100.0%
fabric/membersrvc/ca/ca.go:101:   NewDefaultCertificateSpec 100.0%
fabric/membersrvc/ca/ca.go:121:   GetCommonName             100.0%
fabric/membersrvc/ca/ca.go:127:   GetSerialNumber           100.0%
fabric/membersrvc/ca/ca.go:133:   GetPublicKey              100.0%
fabric/membersrvc/ca/ca.go:139:   GetUsage                  100.0%
fabric/membersrvc/ca/ca.go:145:   GetNotBefore              100.0%
fabric/membersrvc/ca/ca.go:151:   GetNotAfter               100.0%
fabric/membersrvc/ca/ca.go:157:   GetOrganization           100.0%
fabric/membersrvc/ca/ca.go:163:   GetCountry                100.0%
fabric/membersrvc/ca/ca.go:169:   GetSubjectKeyID           100.0%
fabric/membersrvc/ca/ca.go:175:   GetSignatureAlgorithm     100.0%
fabric/membersrvc/ca/ca.go:181:   GetExtensions             100.0%
...

```
## Action Items

Write Golang unit tests to cover the newly added Golang functions:

```
func (ca *CA) canRegister(registrar string, newMemberRole string, newMemberMetadataStr string) error
func newMemberMetadata(metadata string) (*MemberMetadata, error)
func (mm *MemberMetadata) canRegister(registrar string, newRole string, newMemberMetadata *MemberMetadata) error
func strsContained(strs1 []string, strs2 []string, registrar string, field string) error
func strContained(str string, strs []string) bool
func role2String(role int) string
func removeQuotes(str string) string 
```

Hi @JonathanLevi,
I was ready with some unit test code for eca but for some reason I'm not able to run any unit test at all :( .
I'm getting the following error when try to invoke " go test eca_test.go"
eca_test.go:30:2: cannot find package "command-/vendor/github.com/golang/protobuf/proto" in any of:
        /opt/go/src/command-/vendor/github.com/golang/protobuf/proto (from $GOROOT)
        /opt/gopath/src/command-/vendor/github.com/golang/protobuf/proto (from $GOPATH)
FAIL    command-line-arguments [setup failed]

This used to work for me earlier but once i updated my fork with the latest changes from main repo I'm seeing this error. Any ideas why I'm seeing this error?
I have also put up same question in stackoverflow to see if somebody can help me
http://stackoverflow.com/questions/38138904/not-able-to-run-golang-unit-tests-inside-membersrvc-ca

Thanks,
Sudeepta

--- a/membersrvc/ca/eca_test.go
+++ b/membersrvc/ca/eca_test.go
@@ -35,12 +35,14 @@ import (
 )
 
 type User struct {
-	enrollID        string
-	enrollPwd       []byte
-	enrollPrivKey   *ecdsa.PrivateKey
-	role            int
-	affiliation     string
-	affiliationRole string
+	enrollID               string
+	enrollPwd              []byte
+	enrollPrivKey          *ecdsa.PrivateKey
+	role                   int
+	affiliation            string
+	affiliationRole        string
+	registrarRoles         []string
+	registrarDelegateRoles []string
 }
 
 var (
@@ -49,6 +51,14 @@ var (
 	testUser    = User{enrollID: "testUser", role: 1, affiliation: "institution_a", affiliationRole: "00001"}
 	testUser2   = User{enrollID: "testUser2", role: 1, affiliation: "institution_a", affiliationRole: "00001"}
 	testAuditor = User{enrollID: "testAuditor", role: 8}
+	testClient1 = User{enrollID: "testClient1", role: 1, affiliation: "institution_a", affiliationRole: "00001",
+		registrarRoles: []string{"client"}, registrarDelegateRoles: []string{"client"}}
+	testClient2 = User{enrollID: "testClient2", role: 1, affiliation: "institution_a", affiliationRole: "00001",
+		registrarRoles: []string{"client"}}
+	testClient3 = User{enrollID: "testClient2", role: 1, affiliation: "institution_a", affiliationRole: "00001",
+		registrarRoles: []string{"client"}}
+	testPeer = User{enrollID: "testPeer", role: 2, affiliation: "institution_a", affiliationRole: "00001",
+		registrarRoles: []string{"peer"}}
 )
 
 //helper function for multiple tests
@@ -160,8 +170,12 @@ func registerUser(registrar User, user *User) error {
 		Role:        pb.Role(user.role),
 		Account:     user.affiliation,
 		Affiliation: user.affiliationRole,
-		Registrar:   &pb.Registrar{Id: &pb.Identity{Id: registrar.enrollID}},
-		Sig:         nil}
+		Registrar: &pb.Registrar{
+			Id:            &pb.Identity{Id: registrar.enrollID},
+			Roles:         user.registrarRoles,
+			DelegateRoles: user.registrarDelegateRoles,
+		},
+		Sig: nil}
 
 	//sign the req
 	hash := primitives.NewHash()
@@ -286,6 +300,70 @@ func TestRegisterUserNonRegistrar(t *testing.T) {
 	if err == nil {
 		t.Fatal("User without registrar metadata should not be able to register a new user")
 	}
+	t.Logf("Expected an error and indeed received: [%s]", err.Error())
+}
+
+//testAdmin should NOT be able to register testPeer since testAdmin's
+//delegateRoles field DOES NOT contain the value "peer"
+func TestRegisterUserPeer(t *testing.T) {
+
+	err := registerUser(testAdmin, &testPeer)
+
+	if err == nil {
+		t.Fatal("User without appropriate delegateRoles should not be able to register a new user")
+	}
+	t.Logf("Expected an error and indeed received: [%s]", err.Error())
+}
+
+//testAdmin should be able to register testClient1 since testAdmin's
+//delegateRoles field contains the value "client"
+func TestRegisterUserClient(t *testing.T) {
+
+	err := registerUser(testAdmin, &testClient1)
+
+	if err != nil {
+		t.Error(err.Error())
+	}
+}
+
+//testClient1 registered in the previous test should be able to enroll
+func TestCreateCertificatePairClient(t *testing.T) {
+
+	err := enrollUser(&testClient1)
+
+	if err != nil {
+		t.Fatalf("Failed to enroll testClient1: [%s]", err.Error())
+	}
+}
+
+//testClient1 should be able to register testClient2 since testClient1's
+//delegateRoles field contains the value "client"
+func TestRegisterUserClientAsRegistrar(t *testing.T) {
+
+	err := registerUser(testClient1, &testClient2)
+
+	if err != nil {
+		t.Error(err.Error())
+	}
+
+}
+
+//testClient2 should NOT be able to register testClient3 since testClient2's
+//delegateRoles field is empty
+func TestRegisterUserNoDelegateRoles(t *testing.T) {
+
+	err := enrollUser(&testClient2)
+
+	if err != nil {
+		t.Fatalf("Failed to enroll testClient2: [%s]", err.Error())
+	}
+
+	err = registerUser(testClient2, &testClient3)
+
+	if err == nil {
+		t.Fatal("User without delegateRoles should not be able to register a new user")
+	}
+
 	t.Logf("Expected an error and indeed received: [%s]", err.Error())
 }
 

